### DISCUSSION TITLE
Database Entity Selection: Case study on SimssaDB and CantusDB, and Application to thesession.org

### DISCUSSION THREAD

**Discussion by malajvan** [DDMAL/linkedmusic-datalake#12] [General]
@annamorphism, @jacobdgm, and I had an interesting discussion on selecting the primary entity for each database.

In the case of **SimssaDB**, we reached a consensus to designate "musical_work_id" as the primary entity. This entity encompasses a collection of files, each having its own "file_id." The rationale behind this choice lies in the fact that the files within Simssadb essentially represent different formats of the same musical work.

Conversely, when deliberating on **CantusDB**, we opted for "chant_id" as the main entity instead of "cantusID" (which represents text). The reason was that each chant signifies a distinct "version" of the text.

From these two examples, we derived valuable criteria for selecting an appropriate entity within a database:

**1. Data Content:** The purpose served by a database's entities hinges on the nature of the collected data and its real-world relationships.
**2. Database Specification:** Related to the database's design, encompassing factors such as table relationships and the central merging identifier, if applicable.

We then extended our analysis to thesession.org's data. In consultation with Anna and Jacob, we established that each "setting" represents a transcription of variations for each tune. Drawing a parallel to how chants relate to texts in CantusDB, we determined that, by the first criterion, each setting should be regarded as an entity when generating JSON-LD instead of tune as previously implemented.

However, upon conducting a more detailed examination of the database, we observed that the remaining tables we wish to extract information from, such as 'recordings,' 'aliases,' and 'popularity,' are all linked to the 'tunes' table through the 'tune_id' column, rather than to the "setting_id." Consequently, the second criterion compels us to consider each "tune" as the primary entity. This decision aligns with the database's design, mitigates redundancy, and ensures efficient data flattening.

You can see both examples in my fork after I generated the corresponding json files. [settings_naive.json](https://github.com/malajvan/linkedmusic-datalake/blob/main/the_session/jsonld/settings_naive.json) and [tunes_naive.json](https://github.com/malajvan/linkedmusic-datalake/blob/main/the_session/jsonld/tunes_naive.json)
Here is an example for a single tune with 2 recordings and 2 settings. Note that the number of recordings and settings can go up to 200 each.

**a. from `settings_naive.json`:**
```
   {
        "tune_id": 2711,
        "popularity_tunebooks": 332,
        "tunes_setting_id": 15944,
        "tunes_name": "250 To Vigo",
        "tunes_type": "reel",
        "tunes_meter": "4/4",
        "tunes_mode": "Bminor",
        "tunes_date": "2008-05-21 08:10:54",
        "tunes_username": "DonaldK",
        "aliases_alias": "['2:50 To Vigo', '250cc To Vigo', 'Two Fifty To Vigi', 'Two Fifty To Vigo', 'Two-Fifty To Vigo', 'Two-Fifty To Vigo, The']",
        "recordings": [
            {
                "recordings_id": 3815,
                "recordings_artist": "Brongaene Griffin With Gerry O'Beirne And Kevin Burke",
                "recordings_recording": "Three Colours Ginger",
                "recordings_track": 12,
                "recordings_number": 1
            },
            {
                "recordings_id": 2997,
                "recordings_artist": "Duncan Chisholm",
                "recordings_recording": "Farrar",
                "recordings_track": 8,
                "recordings_number": 1
            },
           ...
        ]
    },
    {
        "tune_id": 2711,
        "popularity_tunebooks": 332,
        "tunes_setting_id": 15945,
        "tunes_name": "250 To Vigo",
        "tunes_type": "reel",
        "tunes_meter": "4/4",
        "tunes_mode": "Bminor",
        "tunes_date": "2011-03-31 05:56:50",
        "tunes_username": "Edgar Bolton",
        "aliases_alias": "['2:50 To Vigo', '250cc To Vigo', 'Two Fifty To Vigi', 'Two Fifty To Vigo', 'Two-Fifty To Vigo', 'Two-Fifty To Vigo, The']",
        "recordings": [
            {
                "recordings_id": 3815,
                "recordings_artist": "Brongaene Griffin With Gerry O'Beirne And Kevin Burke",
                "recordings_recording": "Three Colours Ginger",
                "recordings_track": 12,
                "recordings_number": 1
            },
            {
                "recordings_id": 2997,
                "recordings_artist": "Duncan Chisholm",
                "recordings_recording": "Farrar",
                "recordings_track": 8,
                "recordings_number": 1
            }...
        ]
    } 
keeps going for all settings of tune_id 2711
```
   

**b. from `tunes_naive.json`**
```
{
        "tune_id": 102,
        "popularity_tunebooks": 143,
        "tunes_name": "Colonel McBain",
        "aliases_alias": "['Brian Boru', \"Captain MacBean's\", \"Captain McBane's\", 'Col McBain', \"Dan Sullivan's\", 'Devonshire, The', 'Duke Of Clarence, The', 'General McBean', 'John Frank', \"Johnny's Wedding\", 'Sean Frank', \"Sean Frank's\", 'Smoky House, The', 'Sporting Mollie']",
        "recordings": [
            {
                "recordings_id": 887,
                "recordings_artist": "Angelina Carberry And Martin Quinn",
                "recordings_recording": "Angelina Carberry And Martin Quinn",
                "recordings_track": 9,
                "recordings_number": 1
            },
            {
                "recordings_id": 2632,
                "recordings_artist": "Colin Grant",
                "recordings_recording": "Colin Grant",
                "recordings_track": 9,
                "recordings_number": 4
            },
              .....
        ],
        "tune_settings": [
            {
                "tunes_setting_id": 102,
                "tunes_type": "reel",
                "tunes_meter": "4/4",
                "tunes_mode": "Eminor",
                "tunes_date": "2001-05-25 04:05:35",
                "tunes_username": "Jeremy"
            },
            {
                "tunes_setting_id": 4793,
                "tunes_type": "reel",
                "tunes_meter": "4/4",
                "tunes_mode": "Gmajor",
                "tunes_date": "2005-07-28 16:10:44",
                "tunes_username": "dafydd"
            }
                     .....
        ]
    }
```

**Comment by annamorphism** [comment]
I came to a similar conclusion about recordings after our discussion ended yesterday! 
(side note: it seems weird to me that "tunes_type" seems to live on each individual setting--is there ever a case where one tune has e.g. both "reel" and "strathspey" settings? What determines the tune's category in the database?)
Just to ease confusion it might be nice to label which things belong to the tune vs. the tune setting--e.g. I imagine the tunes themselves are entered by individual people on some date, and then each setting is added by other users on some other date.

**Reply by malajvan** [reply]
To approach your questions:
1. I have not checked if one tune has both "reel" and "strathspey" setting type (or anything of the same nature). Will update.

2. For the second paragraph, I would like to add that except for tune_id (which is used to merge every table together), every column's name is modified so that the name of the table is a prefix ie. 'tunes_name', 'tunes_mode'.. comes from `tunes.csv`, `recordings_number`, `recordings_id` comes from `recordings.csv` and so on. (I would add this information to the main discussion.  3. 

     Now to answer your question, the table `tunes` (tunes.csv) doesn't specify which is a setting label and which is a tune label. Every row is distinct by the primary key pair (tune_id, setting_id) and not either one of them. There is no `settings.csv`.
Therefore during my flattening and regrouping, I decided to merge settings by tune_id and tunes_name since the names are always the same, but there was no official indication that tunes_name is a label of tunes and the rest (tunes_meter, tunes_mode,...) are labels of individual settings.

**Reply by malajvan** [reply]
@annamorphism I checked and I'm confident that the tunes_type is a tune labels (as in all settings of the same tune should have the same type). I will fix the json-ld to reflect this, thank you!

**Reply by malajvan** [reply]
@annamorphism a bit late but i've switch tune_type to a tune label instead of a settings' label
